README.md
# EscapeTheRoom
 First project I've ever made. The official name is Space Travel.


changelog.txt
Newest Version: 1.1.0

1.1.0:
Added functionality to info button
Minor changes in tutorial.py'
Version display is now read from this file

Levels/endless.py
import pygame
import os
from Levels import game_classes
from Levels import game_mini_screen
import read_high_score as high_score

pygame.init()

# Functions
def rotate_center(img, angle, x, y):
    pass

def bullet_sphere_blit(length):
    if length == 0:
        SCREEN.blit(bullet_sphere_2, (200, 5))
        return None
    if length == 1:
        SCREEN.blit(bullet_sphere_1, (200, 5))
        return None
    if length == 2:
        SCREEN.blit(bullet_sphere_0, (200, 5))
        return None
    SCREEN.blit(bullet_sphere_reload, (200, 5))


# Variables
QUIT = None
WIDTH, HEIGHT = 1000, 600
res = (WIDTH, HEIGHT)
SCREEN = pygame.display.set_mode(res)

# pygame.image.load(os.path.join("Textures/", ""))

# Load background images
bg = pygame.image.load(os.path.join("Textures/Backgrounds", "level_bg_2.png"))
# Load graphics images
icon = pygame.image.load(os.path.join("Textures/Graphics", "icon.png"))
bullet_sphere_0 = pygame.image.load(os.path.join("Textures/Graphics", "bullet_sphere_0.png"))
bullet_sphere_1 = pygame.image.load(os.path.join("Textures/Graphics", "bullet_sphere_1.png"))
bullet_sphere_2 = pygame.image.load(os.path.join("Textures/Graphics", "bullet_sphere_2.png"))
bullet_sphere_reload = pygame.image.load(os.path.join("Textures/Graphics", "bullet_sphere_reload.png"))
# Load object images
player_orb_img = pygame.image.load(os.path.join("Textures/Objects", "player_orb.png"))   # player_orb
shield_orb_img = pygame.image.load(os.path.join("Textures/Objects", "shield_orb.png"))
bullet_img = pygame.image.load(os.path.join("Textures/Objects", "bullet_1_red.png"))
stable_turret_img = pygame.image.load(os.path.join("Textures/Objects", "turret_stable.png"))
block_breakable_img = pygame.image.load(os.path.join("Textures/Objects", "block_breakable.png"))
block_unbreakable_img = pygame.image.load(os.path.join("Textures/Objects", "block_unbreakable.png"))
block_shield_img = pygame.image.load(os.path.join("Textures/Objects", "block_shield.png"))
laser_img = pygame.image.load(os.path.join("Textures/Objects", "laser.png"))

SCREEN.get_height()
SCREEN.get_width()
def endless_lvl():
    global QUIT
    QUIT = False
    # Instantiate objects
    player = game_classes.Player(30, 285, player_orb_img, shield_orb_img, bullet_img)   # Main player
    breakable_block = game_classes.BlockBreakable(block_breakable_img)
    unbreakable_block = game_classes.BlockUnbreakable(block_unbreakable_img)
    powerup = game_classes.Powerup(block_shield_img, laser_img)
    game_function = game_classes.GameFunction()
    global_var = game_classes.GlobalVar()

    # Variables
    player.shield_count = 1
    font = pygame.font.SysFont("Calibri", 24, italic=True)
    FPS = 60
    fpsClock = pygame.time.Clock()
    # Display
    SCREEN.blit(player.img, (player.pos_x, player.pos_y))
    # In main() Variables
    color = (255, 255, 12)
    isOn = True
    # Move
    direction = 1
    #check = 0
    # Game loop
    while isOn:
        #if check < 10:  check += 1
        #else:   break

        # SCREEN displays
        SCREEN.blit(bg, (0, 0))
        event_list = pygame.event.get()
        for ev in event_list:   # Events check
            if ev.type == pygame.QUIT:
                isOn = False
                QUIT = True
            if ev.type == pygame.USEREVENT:
                player.reload_sec -= 1
            if ev.type == pygame.KEYDOWN:
                if ev.key == pygame.K_d:
                    player.bullet_gen()
                if ev.key == pygame.K_f:
                    direction = direction*(-1)
                if ev.key == pygame.K_ESCAPE:       # ESCAPE
                    is_exit = ecs_screen()
                    if is_exit:
                        isOn = False
        # End event check
        display_score = font.render(f"Score: {player.score}", True, color)
        display_reload = font.render(str(player.reload_sec), True, color)
        display_shield = powerup.font.render("Shield(s): " + str(player.shield_count), True, color)
        SCREEN.blit(display_score, (5, 5))
        SCREEN.blit(display_reload, (209, 45))
        SCREEN.blit(display_shield, (WIDTH - 150, 5))
        # Bullet timer
        if player.bullet[0][0] == player.max_bullet:       # Start timer
            player.bullet[0][0] = player.max_bullet + 10    # A number > max_bullet
            # player.bullet[0][0] = 0
            pygame.time.set_timer(pygame.USEREVENT, 1000)   # 1000ms = 1s
        if player.reload_sec <= 0:       # Reset timer
            player.reload_sec = player.reload_sec_reset
            player.bullet[0][0] = 0
            pygame.time.set_timer(pygame.USEREVENT, 0)
        move_condition = ((player.pos_y + direction*player.vel + player.img.get_height() < HEIGHT) and
                          (player.pos_y + direction*player.vel + 1 > 0))
        if move_condition:
            player.pos_y += direction * player.vel

        bullet_sphere_blit(player.bullet[0][0])

        # Block gen
        unbreakable_block.block_gen_single(gen_type=unbreakable_block.block_tag["gen_type"],
                                           player_pos=(player.pos_x, player.pos_y))

        breakable_block.block_gen_single(gen_type=breakable_block.block_tag["gen_type"],
                                         player_pos=(player.pos_x, player.pos_y))

        powerup.gen_shield(player)

        game_function.block_overlap(breakable_block, unbreakable_block)
        game_function.block_overlap(unbreakable_block, powerup)

        # Block blit
        unbreakable_block.block_blit(SCREEN, player)

        breakable_block.block_blit(SCREEN, player)

        powerup.powerup_blit(SCREEN, player, power_type="shield")

        player.bullet_blit(SCREEN)
        if player.player_collide_single(breakable_block):
            if player.shield_count > 0:
                player.shield_count -= 1
            else:
                game_complete_screen(player, game_state="lose")
                isOn = False
        if player.player_collide_single(unbreakable_block):
            if player.shield_count > 0:
                player.shield_count -= 1
            else:
                game_complete_screen(player, game_state="lose")
                isOn = False

        player.update_player(SCREEN)

        # Multiplier
        game_function.var_multiply(player, breakable_block, unbreakable_block, powerup, global_var)

        #if player.score == 30:
        #    isOn = False

        pygame.display.update()
        fpsClock.tick(FPS)


# pygame.image.load(os.path.join("Textures/", ""))
# Load Background
esc_background = pygame.image.load(os.path.join("Textures/Backgrounds", "esc_background.png"))
lvl_button_continue = pygame.image.load(os.path.join("Textures/Graphics", "lvl_button_continue.png"))
lvl_button_exit = pygame.image.load(os.path.join("Textures/Graphics", "lvl_button_exit.png"))


def ecs_screen():
    global QUIT
    FPS = 60
    fpsClock = pygame.time.Clock()
    # Instantiate Objects
    button_continue = game_mini_screen.ContinueButton(383, 258, lvl_button_continue, lvl_button_continue)
    button_exit = game_mini_screen.ExitButton(561, 258, lvl_button_exit, lvl_button_exit)
    isOn = True
    while isOn:
        event_list = pygame.event.get()
        mouse_pos = pygame.mouse.get_pos()
        for ev in event_list:  # Events check
            if ev.type == pygame.QUIT:
                QUIT = True
                return True
            if ev.type == pygame.KEYDOWN:
                if ev.key == pygame.K_ESCAPE:
                    return False
            if ev.type == pygame.MOUSEBUTTONDOWN:  # Check click
                if button_continue.click_continue(mouse_pos):
                    return False
                if button_exit.click_exit(mouse_pos):  # Quit
                    return True
        SCREEN.blit(esc_background, (200, 100))

        button_continue.button_blit(SCREEN, mouse_pos)
        button_exit.button_blit(SCREEN, mouse_pos)

        pygame.display.update()
        fpsClock.tick(FPS)


def game_complete_screen(player, game_state="lose"):
    global QUIT
    FPS = 60
    fpsClock = pygame.time.Clock()
    font = pygame.font.SysFont("Calibri", 60)
    color = (127, 0, 255)
    if game_state == "win":
        display_win = font.render("Level completed", True, color)
    # Instantiate Objects
    button_continue = game_mini_screen.ContinueButton(383, 258, lvl_button_continue, lvl_button_continue)
    button_exit = game_mini_screen.ExitButton(561, 258, lvl_button_exit, lvl_button_exit)
    isOn = True
    while isOn:
        event_list = pygame.event.get()
        mouse_pos = pygame.mouse.get_pos()
        for ev in event_list:  # Events check
            if ev.type == pygame.QUIT:
                QUIT = True
                return True
            if ev.type == pygame.KEYDOWN:
                if ev.key == pygame.K_ESCAPE:
                    return False
            if ev.type == pygame.MOUSEBUTTONDOWN:  # Check click
                if button_continue.click_continue(mouse_pos):
                    return False
                if button_exit.click_exit(mouse_pos):  # Quit
                    return True
        SCREEN.blit(esc_background, (200, 100))
        if game_state == "lose":
            high_score_num = high_score.read_file()
            if int(high_score_num) < player.score:
                high_score.write_file(player.score)
            display_lose = font.render("Score: " + str(player.score), True, color)
            SCREEN.blit(display_lose, (380, 150))

        button_continue.button_blit(SCREEN, mouse_pos)
        button_exit.button_blit(SCREEN, mouse_pos)

        pygame.display.update()
        fpsClock.tick(FPS)



Levels/tutorial.py
import pygame
import os
from Levels import game_classes
from Levels import game_mini_screen

pygame.init()

# Functions
def rotate_center(img, angle, x, y):
    pass

def bullet_sphere_blit(length):
    if length == 0:
        SCREEN.blit(bullet_sphere_2, (200, 5))
        return None
    if length == 1:
        SCREEN.blit(bullet_sphere_1, (200, 5))
        return None
    if length == 2:
        SCREEN.blit(bullet_sphere_0, (200, 5))
        return None
    SCREEN.blit(bullet_sphere_reload, (200, 5))


# Variables
QUIT = None
WIDTH, HEIGHT = 1000, 600
res = (WIDTH, HEIGHT)
SCREEN = pygame.display.set_mode(res)

# pygame.image.load(os.path.join("Textures/", ""))

# Load background images
bg = pygame.image.load(os.path.join("Textures/Backgrounds", "level_bg_2.png"))
# Load graphics images
icon = pygame.image.load(os.path.join("Textures/Graphics", "icon.png"))
bullet_sphere_0 = pygame.image.load(os.path.join("Textures/Graphics", "bullet_sphere_0.png"))
bullet_sphere_1 = pygame.image.load(os.path.join("Textures/Graphics", "bullet_sphere_1.png"))
bullet_sphere_2 = pygame.image.load(os.path.join("Textures/Graphics", "bullet_sphere_2.png"))
bullet_sphere_reload = pygame.image.load(os.path.join("Textures/Graphics", "bullet_sphere_reload.png"))
# Load object images
player_orb_img = pygame.image.load(os.path.join("Textures/Objects", "player_orb.png"))   # player_orb
shield_orb_img = pygame.image.load(os.path.join("Textures/Objects", "shield_orb.png"))
bullet_img = pygame.image.load(os.path.join("Textures/Objects", "bullet_1_red.png"))
stable_turret_img = pygame.image.load(os.path.join("Textures/Objects", "turret_stable.png"))
block_breakable_img = pygame.image.load(os.path.join("Textures/Objects", "block_breakable.png"))
block_unbreakable_img = pygame.image.load(os.path.join("Textures/Objects", "block_unbreakable.png"))
block_shield_img = pygame.image.load(os.path.join("Textures/Objects", "block_shield.png"))
laser_img = pygame.image.load(os.path.join("Textures/Objects", "laser.png"))

SCREEN.get_height()
SCREEN.get_width()
def tutorial_lvl():
    global QUIT
    QUIT = False
    # Instantiate objects
    player = game_classes.Player(30, 285, player_orb_img, shield_orb_img, bullet_img)   # Main player
    breakable_block = game_classes.BlockBreakable(block_breakable_img)
    unbreakable_block = game_classes.BlockUnbreakable(block_unbreakable_img)
    powerup = game_classes.Powerup(block_shield_img, laser_img)
    game_function = game_classes.GameFunction()
    global_var = game_classes.GlobalVar()

    # Variables
    font = pygame.font.SysFont("Calibri", 24, italic=True)
    FPS = 60
    fpsClock = pygame.time.Clock()
    # Display
    SCREEN.blit(player.img, (player.pos_x, player.pos_y))
    # In main() Variables
    color = (255, 255, 12)
    isOn = True
    # Move
    direction = 1
    #check = 0
    # Game loop
    while isOn:
        #if check < 10:  check += 1
        #else:   break

        # SCREEN displays
        SCREEN.blit(bg, (0, 0))
        event_list = pygame.event.get()
        for ev in event_list:   # Events check
            if ev.type == pygame.QUIT:
                isOn = False
                QUIT = True
            if ev.type == pygame.USEREVENT:
                player.reload_sec -= 1
            if ev.type == pygame.KEYDOWN:
                if ev.key == pygame.K_d:
                    player.bullet_gen()
                if ev.key == pygame.K_f:
                    direction = direction*(-1)
                if ev.key == pygame.K_ESCAPE:       # ESCAPE
                    is_exit = ecs_screen()
                    if is_exit:
                        isOn = False
        # End event check
        display_score = font.render(f"Score: {player.score}", True, color)
        display_reload = font.render(str(player.reload_sec), True, color)
        display_shield = powerup.font.render("Shield(s): " + str(player.shield_count), True, color)
        SCREEN.blit(display_score, (5, 5))
        SCREEN.blit(display_reload, (209, 45))
        SCREEN.blit(display_shield, (WIDTH - 150, 5))
        # Bullet timer
        if player.bullet[0][0] == player.max_bullet:       # Start timer
            player.bullet[0][0] = player.max_bullet + 10    # A number > max_bullet
            # player.bullet[0][0] = 0
            pygame.time.set_timer(pygame.USEREVENT, 1000)   # 1000ms = 1s
        if player.reload_sec <= 0:       # Reset timer
            player.reload_sec = player.reload_sec_reset
            player.bullet[0][0] = 0
            pygame.time.set_timer(pygame.USEREVENT, 0)
        move_condition = ((player.pos_y + direction*player.vel + player.img.get_height() < HEIGHT) and
                          (player.pos_y + direction*player.vel + 1 > 0))
        if move_condition:
            player.pos_y += direction * player.vel

        bullet_sphere_blit(player.bullet[0][0])

        # Block gen
        unbreakable_block.block_gen_single(gen_type=unbreakable_block.block_tag["gen_type"],
                                           player_pos=(player.pos_x, player.pos_y))

        breakable_block.block_gen_single(gen_type=breakable_block.block_tag["gen_type"],
                                         player_pos=(player.pos_x, player.pos_y))

        powerup.gen_shield(player)

        game_function.block_overlap(breakable_block, unbreakable_block)
        game_function.block_overlap(unbreakable_block, powerup)

        # Block blit
        unbreakable_block.block_blit(SCREEN, player)

        breakable_block.block_blit(SCREEN, player)

        powerup.powerup_blit(SCREEN, player, power_type="shield")

        player.bullet_blit(SCREEN)
        if player.player_collide_single(breakable_block):
            if player.shield_count > 0:
                player.shield_count -= 1
            else:
                is_exit = game_complete_screen(game_state="lose")
                if is_exit:
                    isOn = False
        if player.player_collide_single(unbreakable_block):
            if player.shield_count > 0:
                player.shield_count -= 1
            else:
                is_exit = game_complete_screen(game_state="lose")
                if is_exit:
                    isOn = False
        if player.score > 30:
            is_exit = game_complete_screen(game_state="win")
            if is_exit:
                isOn = False

        player.update_player(SCREEN)



        # Multiplier
        game_function.var_multiply(player, breakable_block, unbreakable_block, powerup, global_var)

        #if player.score == 30:
        #    isOn = False

        pygame.display.update()
        fpsClock.tick(FPS)


# pygame.image.load(os.path.join("Textures/", ""))
# Load Background
esc_background = pygame.image.load(os.path.join("Textures/Backgrounds", "esc_background.png"))
lvl_button_continue = pygame.image.load(os.path.join("Textures/Graphics", "lvl_button_continue.png"))
lvl_button_exit = pygame.image.load(os.path.join("Textures/Graphics", "lvl_button_exit.png"))
lvl_button_replay = pygame.image.load(os.path.join("Textures/Graphics", "lvl_button_replay.png"))
def ecs_screen():
    """Small note: the img 'blit' mechanism used here is different from
    the one used in 'endless'. However, they work almost the same and
    serve the same purpose."""
    global QUIT
    FPS = 60
    fpsClock = pygame.time.Clock()
    # Instantiate Objects
    button_continue = game_mini_screen.ContinueButton(383, 258, lvl_button_continue, lvl_button_continue)
    button_exit = game_mini_screen.ExitButton(561, 258, lvl_button_exit, lvl_button_exit)
    isOn = True
    # This part is different in 'endless'
    SCREEN.blit(esc_background, (200, 100))
    button_continue.simple_button_blit(SCREEN)
    button_exit.simple_button_blit(SCREEN)
    pygame.display.update()
    #
    while isOn:
        event_list = pygame.event.get()
        mouse_pos = pygame.mouse.get_pos()
        for ev in event_list:  # Events check
            if ev.type == pygame.QUIT:
                QUIT = True
                return True
            if ev.type == pygame.KEYDOWN:
                if ev.key == pygame.K_ESCAPE:
                    return False
            if ev.type == pygame.MOUSEBUTTONDOWN:  # Check click
                if button_continue.click_continue(mouse_pos):
                    return False
                if button_exit.click_exit(mouse_pos):  # Quit
                    return True

        fpsClock.tick(FPS)


def game_complete_screen(game_state="lose"):
    global QUIT
    FPS = 60
    fpsClock = pygame.time.Clock()
    font = pygame.font.SysFont("Calibri", 60)
    color = (127, 0, 255)
    # Instantiate Objects
    button_continue = game_mini_screen.ContinueButton(383, 258, lvl_button_continue, lvl_button_continue)
    button_exit = game_mini_screen.ExitButton(561, 258, lvl_button_exit, lvl_button_exit)
    button_replay = game_mini_screen.ReplayButton(472, 258, lvl_button_replay, lvl_button_replay)
    isOn = True
    while isOn:
        event_list = pygame.event.get()
        mouse_pos = pygame.mouse.get_pos()
        for ev in event_list:  # Events check
            if ev.type == pygame.QUIT:
                QUIT = True
                return True
            if ev.type == pygame.KEYDOWN:
                if ev.key == pygame.K_ESCAPE:
                    return False
            if ev.type == pygame.MOUSEBUTTONDOWN:  # Check click
                if button_continue.click_continue(mouse_pos):
                    return True
                if button_exit.click_exit(mouse_pos):  # Quit
                    return True
        SCREEN.blit(esc_background, (200, 100))
        if game_state == "lose":
            display_lose = font.render("Game over", True, color)
            SCREEN.blit(display_lose, (370, 150))
        if game_state == "win":
            display_win = font.render("Level completed", True, color)
            SCREEN.blit(display_win, (310, 150))

        button_continue.button_blit(SCREEN, mouse_pos)
        button_exit.button_blit(SCREEN, mouse_pos)

        pygame.display.update()
        fpsClock.tick(FPS)


menu_screen.py
import pygame
import os
import read_high_score as high_score
from Levels import tutorial, endless

pygame.init()

QUIT = True     # QUIT #


class Button:
    def __init__(self, x, y, img, img2):
        self.x = x
        self.y = y
        self.img = img
        self.img2 = img2

    def simple_button_blit(self, screen):
        screen.blit(self.img, (self.x, self.y))

    def button_blit(self, screen, mouse_pos):
        if self.isAbove(mouse_pos):
            screen.blit(self.img2, (self.x, self.y))
        else:
            screen.blit(self.img, (self.x, self.y))

    def isAbove(self, mouse_pos):
        """Return True if the mouse is at the image, else return False"""
        condition = ((self.x < mouse_pos[0] < self.x + self.img.get_width())
                    and (self.y < mouse_pos[1] < self.y + self.img.get_height()))
        return condition


class Button2(Button):
    def __init__(self, x, y, img, img2, moveType):
        super().__init__(x, y, img, img2)
        self.moveType = (1 if moveType == "Forward" else 0)
        self.page = 1  # Used to count pages, starts at 1
        self.max = (5 if self.moveType else 1)

    def button_blit(self, screen, mouse_pos):
        if self.moveType:
            if self.page < self.max:
                if self.isAbove(mouse_pos):
                    screen.blit(self.img2, (self.x, self.y))
                else:
                    screen.blit(self.img, (self.x, self.y))
        else:
            if self.page > self.max:
                if self.isAbove(mouse_pos):
                    screen.blit(self.img2, (self.x, self.y))
                else:
                    screen.blit(self.img, (self.x, self.y))


class ButtonLevel(Button):
    def __init__(self, x, y, img, img2):
        super().__init__(x, y, img, img2)
        self.onClick = False
    def call_tutorial(self, mouse_pos):
        if self.isAbove(mouse_pos):
            tutorial.tutorial_lvl()
            lvl_quit()
    def call_endless(self, mouse_pos):
        if self.isAbove(mouse_pos):
            endless.endless_lvl()
            lvl_quit()


class PageCheck:
    def __init__(self, block_img1, block_img2, block_img3, player_img, player_shield_img):
        self.player_img = player_img
        self.img1 = block_img1
        self.img2 = block_img2
        self.img3 = block_img3
        self.shield_img = player_shield_img
        pass
    def render_text_1(self, screen, x, y):
        temp = font.render("Press F to move up or down. Press D to shoot", True, (0, 0, 0))
        screen.blit(temp, (x, y))
        screen.blit(self.player_img, (483, 284))

    def render_text_2(self, screen, x, y):
        temp = font.render("There are 3 types of block:", True, (0, 0, 0))
        temp2 = font.render("Breakable block, Unbreakable block, Shield Block,", True, (0, 0, 0))
        temp3 = font.render("each will randomly spawn.", True, (0, 0, 0))
        screen.blit(temp, (x, y))
        screen.blit(temp2, (x, y+40))
        screen.blit(temp3, (x, y+80))
        screen.blit(self.img1, (300, 264))
        screen.blit(self.img2, (500, 264))
        screen.blit(self.img3, (700, 275))

    def render_text_0(self, screen, x, y):
        temp = font.render("Each Levels has a number of phases.", True, (0, 0, 0))
        temp2 = font.render("Depends on the Levels, it can have 3, 5, or 7 phases.", True, (0, 0, 0))
        temp3 = font.render("The \"Special Level\" has 22 phases, which are", True, (0, 0, 0))
        temp4 = font.render("the combination of the previous 20 phases and 2 special phases.", True, (0, 0, 0))
        screen.blit(temp, (x, y))
        screen.blit(temp2, (x, y + 40))
        screen.blit(temp3, (x, y + 80))
        screen.blit(temp4, (x, y + 120))

    def render_text_3(self, screen, x, y):
        temp = font.render("Shoot a block to destroy it.", True, (0, 0, 0))
        temp2 = font.render("Unbreakable blocks are indestructible.", True, (0, 0, 0))
        temp3 = font.render("You will lose if you touch a block,", True, (0, 0, 0))
        temp4 = font.render("but Shield block is an exception.", True, (0, 0, 0))
        screen.blit(temp, (x, y))
        screen.blit(temp2, (x, y + 40))
        screen.blit(temp3, (x, y + 80))
        screen.blit(temp4, (x, y + 120))
        screen.blit(self.img3, (483, 284))

    def render_text_4(self, screen, x, y):
        temp = font.render("Destroy a Breakable block increases your score.", True, (0, 0, 0))
        temp2 = font.render("Destroy a Shield block increases you shield count.", True, (0, 0, 0))
        temp3 = font.render("The more shields you have, the more lives you get.", True, (0, 0, 0))
        screen.blit(temp, (x, y))
        screen.blit(temp2, (x, y+40))
        screen.blit(temp3, (x, y+80))
        screen.blit(self.shield_img, (483, 284))

    def render_text_5(self, screen, x, y):
        temp = font.render("Enjoy!", True, (0, 0, 0))
        screen.blit(temp, (x, y))

# Functions
def lvl_quit(no_global=False):
    isQUIT = (tutorial.QUIT or endless.QUIT)
    if not no_global:
        if isQUIT:
            global QUIT
            QUIT = True
    else:
        return isQUIT

# Things
WIDTH, HEIGHT = 1000, 600
MENU_SCREEN = pygame.display.set_mode((WIDTH, HEIGHT))
font = pygame.font.SysFont("Calibri", 25)

# Load Backgrounds
menu_bg = pygame.image.load("Textures/Backgrounds/menu_bg.png")
howtoplay_bg = pygame.image.load("Textures/Backgrounds/howtoplay_bg.png")
# Load Graphics
icon = pygame.image.load(os.path.join("Textures/Graphics", "icon.png"))
button_play_img = pygame.image.load("Textures/Graphics/button_play_1.png")
button_howtoplay_img = pygame.image.load("Textures/Graphics/button_howtoplay_1.png")
button_quit_img = pygame.image.load("Textures/Graphics/button_quit_1.png")
button_info_img = pygame.image.load("Textures/Graphics/button_info_1.png")

pygame.display.set_icon(icon)
pygame.display.set_caption("Space Travel")

# Instantiate Objects
play_button = Button(423, 190, button_play_img, button_play_img)
howtoplay_button = Button(423, 255, button_howtoplay_img, button_howtoplay_img)
info_button = Button(423, 320, button_info_img, button_info_img)
exit_button = Button(423, 385, button_quit_img, button_quit_img)


def intro_main():
    FPS = 60
    clock = pygame.time.Clock()
    # count = 0     # Test variable
    isOn = True
    global QUIT
    QUIT = False
    while isOn:
        # if count < 10:  count +=1
        # else: isOn = False
        event_list = pygame.event.get()
        pos = pygame.mouse.get_pos()
        MENU_SCREEN.blit(menu_bg, (0, 0))
        play_button.button_blit(MENU_SCREEN, pos)
        howtoplay_button.button_blit(MENU_SCREEN, pos)
        info_button.button_blit(MENU_SCREEN, pos)
        exit_button.button_blit(MENU_SCREEN, pos)
        for ev in event_list:  # Events check
            if ev.type == pygame.QUIT:
                isOn = False
                QUIT = True
            if ev.type == pygame.MOUSEBUTTONDOWN:  # Check click
                if play_button.isAbove(pos):  # Play
                    lvl_list()
                    if on_lvl_1.onClick:
                        isOn = False
                    if QUIT:  # if QUIT == True
                        isOn = False
                if howtoplay_button.isAbove(pos):  # How to play
                    howtoplay_main()
                    if QUIT:  # if QUIT == True
                        isOn = False
                if exit_button.isAbove(pos):  # Quit
                    isOn = False
                    QUIT = True
                if info_button.isAbove(pos):
                    show_info()
                    if QUIT:
                        isOn = False

        pygame.display.update()
        clock.tick(FPS)


# Load How-to-play images
player_orb = pygame.image.load("Textures/Objects/player_orb.png")
button_back = pygame.image.load("Textures/Graphics/button_back.png")
button_forward = pygame.image.load("Textures/Graphics/button_forward.png")
button_backward = pygame.image.load("Textures/Graphics/button_backward.png")
breakable_block = pygame.image.load("Textures/Objects/block_breakable.png")
unbreakable_block = pygame.image.load("Textures/Objects/block_unbreakable.png")
shield_block = pygame.image.load("Textures/Objects/block_shield.png")
player_shield = pygame.image.load("Textures/Objects/shield_orb.png")
# Instantiate How-to-play Objects
back_button = Button(15, 15, button_back, button_back)
forward_button = Button2(953, 284, button_forward, button_forward, "Forward")
backward_button = Button2(15, 284, button_backward, button_backward, "Backward")
page_check = PageCheck(breakable_block, unbreakable_block, shield_block, player_orb, player_shield)


def howtoplay_main():
    #count = 0
    # How-to-play Texts
    _move = font.render("Use W, S, A, D to move up, down, left, right", True, (0, 0, 0))
    # _test = font.render(f"{backward_button.moveType}", True, (0, 0, 0))
    FPS = 60
    clock = pygame.time.Clock()
    isOn = True
    global QUIT
    QUIT = False
    while isOn:
        #if count < 10:  count +=1
        #else: isOn = False
        event_list = pygame.event.get()
        pos = pygame.mouse.get_pos()
        for ev in event_list:  # Events check
            if ev.type == pygame.QUIT:
                isOn = False
                QUIT = True
            if ev.type == pygame.MOUSEBUTTONDOWN:  # Check click
                if back_button.isAbove(pos):  # Play
                    isOn = False
                if forward_button.isAbove(pos) and forward_button.page < forward_button.max:
                    forward_button.page += 1
                    backward_button.page += 1
                if backward_button.isAbove(pos) and backward_button.page > backward_button.max:
                    backward_button.page -= 1
                    forward_button.page -= 1
        MENU_SCREEN.blit(howtoplay_bg, (0, 0))
        back_button.button_blit(MENU_SCREEN, pos)
        forward_button.button_blit(MENU_SCREEN, pos)
        backward_button.button_blit(MENU_SCREEN, pos)
        # Render Texts
        enjoy = (430 if backward_button.page == 5 else 253)
        getattr(page_check, "render_text_" + str(backward_button.page), "")(MENU_SCREEN, enjoy, 336)
        pygame.display.update()
        clock.tick(FPS)


# Load lvl-list images
lvl_bg = pygame.image.load("Textures/Backgrounds/lvl_bg.png")
level_1_img = pygame.image.load("Textures/Graphics/lvl_1.png")
level_2_img = pygame.image.load("Textures/Graphics/lvl_2.png")
level_3_img = pygame.image.load("Textures/Graphics/lvl_3.png")
level_4_img = pygame.image.load("Textures/Graphics/lvl_4.png")
level_special_img = pygame.image.load("Textures/Graphics/lvl_special.png")
# Instantiate lvl-list Objects
# Instantiate lvl-list Objects
on_lvl_1 = ButtonLevel(200, 260, level_1_img, level_1_img)
on_endless = ButtonLevel(721, 260, level_special_img, level_special_img)


def lvl_list():
    color = (255, 116, 17)
    text_tutorial = font.render("Tutorial", True, color)
    text_endless = font.render("Endless", True, color)
    text_high_score = font.render("Highest Score: " + high_score.read_file(), True, (0, 0, 0))
    FPS = 60
    clock = pygame.time.Clock()
    global QUIT
    QUIT = False
    isOn = True
    #count = 0
    while isOn:
        #if count < 10:  count +=1
        #else: isOn = False
        event_list = pygame.event.get()
        pos = pygame.mouse.get_pos()
        MENU_SCREEN.blit(lvl_bg, (0, 0))
        back_button.button_blit(MENU_SCREEN, pos)
        on_lvl_1.button_blit(MENU_SCREEN, pos)
        on_endless.button_blit(MENU_SCREEN, pos)
        MENU_SCREEN.blit(text_tutorial, (200, 340))
        MENU_SCREEN.blit(text_endless, (721, 340))
        MENU_SCREEN.blit(text_high_score, (680, 370))
        for ev in event_list:
            if ev.type == pygame.QUIT:
                isOn = False
                QUIT = True
            if ev.type == pygame.MOUSEBUTTONDOWN:  # Check click
                if back_button.isAbove(pos):  # Play
                    isOn = False
                on_lvl_1.call_tutorial(pos)
                # Small note: Quitting Levels midway was handle by lvl_quit
                if lvl_quit(no_global=True):
                    isOn = False

            if ev.type == pygame.MOUSEBUTTONDOWN:  # Check click
                if back_button.isAbove(pos):  # Play
                    isOn = False
                on_endless.call_endless(pos)
                text_high_score = font.render("Highest Score: " + high_score.read_file(), True, (0, 0, 0))
                MENU_SCREEN.blit(text_high_score, (680, 370))
                # Small note: Quitting Levels midway was handle by lvl_quit
                if lvl_quit(no_global=True):
                    isOn = False

        pygame.display.update()
        clock.tick(FPS)


# Load show_info images
info_bg = pygame.image.load("Textures/Backgrounds/info.png")
# Instantiate show_info objects
back_button2 = Button(215, 115, button_back, button_back)


def show_info():
    FPS = 60
    clock = pygame.time.Clock()
    isOn = True
    global QUIT
    QUIT = False
    new_font = pygame.font.SysFont("Calibri", 20)

    with open("log.txt", "r") as version:
        ver_display = version.readline()
    ver_display = ver_display[0:-1]
    ver_display = ver_display[7:]
    print(ver_display)
    ver_display2 = new_font.render(ver_display, True, (0, 0, 0))
    print(ver_display2)

    while isOn:
        # if count < 10:  count +=1
        # else: isOn = False
        event_list = pygame.event.get()
        pos = pygame.mouse.get_pos()
        MENU_SCREEN.blit(info_bg, (200, 100))
        MENU_SCREEN.blit(ver_display2, (215, 470))
        back_button2.button_blit(MENU_SCREEN, pos)
        for ev in event_list:  # Events check
            if ev.type == pygame.QUIT:
                isOn = False
                QUIT = True
            if ev.type == pygame.MOUSEBUTTONDOWN:  # Check click
                if back_button2.isAbove(pos):  # Play
                    isOn = False
        pygame.display.update()
        clock.tick(FPS)


def main():     # Main game loop
    """I do not use 'isOn' for this loop because this is the main game
    loop, the variable 'QUIT' will determine if the game run or not."""
    global QUIT
    QUIT = False
    while not QUIT:
        event_list = pygame.event.get()
        for ev in event_list:
            if ev.type == pygame.QUIT:
                QUIT = True
        intro_main()
